nixos/modules/flyingcircus/packages/fcmanage/fc/manage/resize.py

```diff
@@ -286,10 +291,33 @@ def check_qemu_reboot():
             comment=msg))
 
 
+def kernel_version(kernel):
+    """Guesses kernel version from /run/*-system/kernel.
+
+    Theory of operation: A link like `/run/current-system/kernel` points
+    to a bzImage like `/nix/store/abc...-linux-4.4.27/bzImage`. The
+    directory also contains a `lib/modules` dir which should have the
+    kernel version as sole subdir, e.g.
+    `/nix/store/abc...-linux-4.4.27/lib/modules/4.4.27`. This function
+    returns that version number or bails out if the assumptions laid down here
+    do not hold.
+    """
+    bzImage = os.readlink(kernel)
+    moddir = os.listdir(p.join(p.dirname(bzImage), 'lib', 'modules'))
+    if len(moddir) != 1:
+        raise RuntimeError('modules subdir does not contain exactly '
+                           'one item', moddir)
+    return moddir[0]
+
+
 def check_kernel_reboot():
     """Schedules a reboot if the kernel has changed."""
-    if (os.stat('/run/current-system/kernel').st_ino !=
-            os.stat('/run/booted-system/kernel').st_ino):
+    booted, current = map(kernel_version, [
+        '/run/booted-system/kernel',
+        '/run/current-system/kernel'])
+    verbose('check_kernel: booted={}, current={}'.format(booted, current))
+    if booted != current:
+        print('kernel changed: scheduling reboot')
```

Comment:
if verbose is usually off, then I'd suggest adding booted/current to this message, too